From 0d4690c116a1775e4e129878d9f5a2f5f77c068f Mon Sep 17 00:00:00 2001
From: Jimi Huotari <chiitoo@gentoo.org>
Date: Sun, 23 Dec 2018 23:30:57 +0200
Subject: [PATCH] lxqt-config-input: make touchpad settings/dependencies
 optional

Add a configure option for building the touchpad section of the
'Keyboard and Mouse Settings' (enabled by default).  This allows
for reducing the dependencies on systems that do not have a use
for them.

Closes: https://github.com/lxqt/lxqt-config/issues/358
---
 lxqt-config-input/CMakeLists.txt        | 71 ++++++++++++++++++++-----
 lxqt-config-input/lxqt-config-input.cpp | 10 ++++
 lxqt-config-input/touchpad-config.h.in  |  1 +
 3 files changed, 69 insertions(+), 13 deletions(-)
 create mode 100644 lxqt-config-input/touchpad-config.h.in

diff --git a/lxqt-config-input/CMakeLists.txt b/lxqt-config-input/CMakeLists.txt
index a315e71..8c4c0fb 100644
--- a/lxqt-config-input/CMakeLists.txt
+++ b/lxqt-config-input/CMakeLists.txt
@@ -1,24 +1,44 @@
 project(lxqt-config-input)
 
+option(WITH_TOUCHPAD "Build touchpad settings" ON)
+
 find_package(X11 REQUIRED)
-find_package(PkgConfig REQUIRED)
-pkg_check_modules(XORG_LIBINPUT REQUIRED xorg-libinput)
-pkg_check_modules(XORG_XI REQUIRED xi)
-pkg_check_modules(LIBUDEV REQUIRED libudev)
 
-include_directories(
+if (WITH_TOUCHPAD)
+    find_package(PkgConfig REQUIRED)
+    pkg_check_modules(XORG_LIBINPUT REQUIRED xorg-libinput)
+    pkg_check_modules(XORG_XI REQUIRED xi)
+    pkg_check_modules(LIBUDEV REQUIRED libudev)
+endif ()
+
+set(lxqt-config-input_INCS
     ${X11_INCLUDE_DIR}
-    ${XORG_LIBINPUT_INCLUDE_DIRS}
     "${CMAKE_CURRENT_SOURCE_DIR}/../liblxqt-config-cursor"
 )
 
+if (WITH_TOUCHPAD)
+    set(lxqt-config-input_INCS
+        ${lxqt-config-input_INCS}
+        ${XORG_LIBINPUT_INCLUDE_DIRS}
+    )
+endif ()
+
+include_directories(
+    ${lxqt-config-input_INCS}
+)
+
+configure_file(
+    "${CMAKE_CURRENT_SOURCE_DIR}/touchpad-config.h.in"
+    "${CMAKE_CURRENT_SOURCE_DIR}/touchpad-config.h"
+    @ONLY
+)
+
 set(lxqt-config-input_HDRS
     keyboardconfig.h
     mouseconfig.h
     mouseconfig.h
     keyboardlayoutconfig.h
     selectkeyboardlayoutdialog.h
-    touchpadconfig.h
 )
 
 set(lxqt-config-input_SRCS
@@ -27,8 +47,6 @@ set(lxqt-config-input_SRCS
     mouseconfig.cpp
     keyboardlayoutconfig.cpp
     selectkeyboardlayoutdialog.cpp
-    touchpadconfig.cpp
-    touchpaddevice.cpp
 )
 
 set(lxqt-config-input_UIS
@@ -36,9 +54,26 @@ set(lxqt-config-input_UIS
     keyboardconfig.ui
     keyboardlayoutconfig.ui
     selectkeyboardlayoutdialog.ui
-    touchpadconfig.ui
 )
 
+if (WITH_TOUCHPAD)
+    set(lxqt-config-input_HDRS
+        ${lxqt-config-input_HDRS}
+        touchpadconfig.h
+    )
+
+    set(lxqt-config-input_SRCS
+        ${lxqt-config-input_SRCS}
+        touchpadconfig.cpp
+        touchpaddevice.cpp
+    )
+
+    set(lxqt-config-input_UIS
+        ${lxqt-config-input_UIS}
+        touchpadconfig.ui
+    )
+endif ()
+
 # Translations **********************************
 lxqt_translate_ts(QM_FILES
     UPDATE_TRANSLATIONS
@@ -67,14 +102,24 @@ add_executable(lxqt-config-input
     ${QM_LOADER}
 )
 
-target_link_libraries(lxqt-config-input
+set(lxqt-config-input_TLBS
     Qt5::Widgets
     Qt5::X11Extras
     ${X11_LIBRARIES}
-    ${X11_Xinput_LIB}
     lxqt
     lxqt-config-cursor
-    udev
+)
+
+if (WITH_TOUCHPAD)
+    set(lxqt-config-input_TLBS
+        ${lxqt-config-input_TLBS}
+        ${X11_Xinput_LIB}
+        udev
+    )
+endif ()
+
+target_link_libraries(lxqt-config-input
+    ${lxqt-config-input_TLBS}
 )
 
 set_target_properties(lxqt-config-input
diff --git a/lxqt-config-input/lxqt-config-input.cpp b/lxqt-config-input/lxqt-config-input.cpp
index 9aa3134..ec7fa73 100644
--- a/lxqt-config-input/lxqt-config-input.cpp
+++ b/lxqt-config-input/lxqt-config-input.cpp
@@ -25,8 +25,12 @@
 #include "keyboardconfig.h"
 #include "../liblxqt-config-cursor/selectwnd.h"
 #include "keyboardlayoutconfig.h"
+#include "touchpad-config.h"
+
+#ifdef WITH_TOUCHPAD
 #include "touchpadconfig.h"
 #include "touchpaddevice.h"
+#endif
 
 int main(int argc, char** argv) {
     LXQt::SingleApplication app(argc, argv);
@@ -41,9 +45,11 @@ int main(int argc, char** argv) {
     app.setApplicationVersion(VERINFO);
 
     dlgOptions.setCommandLine(&parser);
+#ifdef WITH_TOUCHPAD
     QCommandLineOption loadOption("load-touchpad",
             app.tr("Load last touchpad settings."));
     parser.addOption(loadOption);
+#endif
     parser.addVersionOption();
     parser.addHelpOption();
     parser.process(app);
@@ -54,11 +60,13 @@ int main(int argc, char** argv) {
       configName = "session";
     LXQt::Settings settings(configName);
 
+#ifdef WITH_TOUCHPAD
     bool loadLastTouchpadSettings = parser.isSet(loadOption);
     if (loadLastTouchpadSettings) {
         TouchpadDevice::loadSettings(&settings);
         return 0;
     }
+#endif
 
     LXQt::ConfigDialog dlg(QObject::tr("Keyboard and Mouse Settings"), &settings);
     app.setActivationWindow(&dlg);
@@ -80,10 +88,12 @@ int main(int argc, char** argv) {
     dlg.addPage(keyboardLayoutConfig, QObject::tr("Keyboard Layout"), "input-keyboard");
     QObject::connect(&dlg, SIGNAL(reset()), keyboardLayoutConfig, SLOT(reset()));
 
+#ifdef WITH_TOUCHPAD
     TouchpadConfig* touchpadConfig = new TouchpadConfig(&settings, &dlg);
     dlg.addPage(touchpadConfig, QObject::tr("Mouse and Touchpad"), "input-tablet");
     QObject::connect(&dlg, &LXQt::ConfigDialog::reset,
                      touchpadConfig, &TouchpadConfig::reset);
+#endif
 
     dlg.setWindowIcon(QIcon::fromTheme("input-keyboard"));
 
diff --git a/lxqt-config-input/touchpad-config.h.in b/lxqt-config-input/touchpad-config.h.in
new file mode 100644
index 0000000..d5bdab6
--- /dev/null
+++ b/lxqt-config-input/touchpad-config.h.in
@@ -0,0 +1 @@
+#cmakedefine WITH_TOUCHPAD @WITH_TOUCHPAD@
-- 
2.20.1

